From e1ad0c3b31fe0f5a0d9676e7901bc8d175ce26e6 Mon Sep 17 00:00:00 2001
From: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
Date: Mon, 2 Mar 2015 17:40:35 +0100
Subject: [PATCH] wifi: platform device for iwlwifi

Add iwlwifi platform header file
managing platform specific resources for iwlwifi driver

Change-Id: I5d11b5acb240111f7434a91f2f17398af2b9f0bf
Signed-off-by: Guillaume Aubertin <guillaume.aubertin@intel.com>
Signed-off-by: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
---
 arch/x86/platform/intel-mid/device_libs/Makefile   |   1 +
 .../intel-mid/device_libs/platform_iwlwifi.c       | 108 +++++++++++++++++++++
 drivers/acpi/acpi_platform.c                       |   2 +
 include/linux/platform_data/iwlwifi.h              |  35 +++++++
 4 files changed, 146 insertions(+)
 create mode 100644 arch/x86/platform/intel-mid/device_libs/platform_iwlwifi.c
 create mode 100644 include/linux/platform_data/iwlwifi.h

diff --git a/arch/x86/platform/intel-mid/device_libs/Makefile b/arch/x86/platform/intel-mid/device_libs/Makefile
index b05c054..17decf9 100644
--- a/arch/x86/platform/intel-mid/device_libs/Makefile
+++ b/arch/x86/platform/intel-mid/device_libs/Makefile
@@ -23,3 +23,4 @@ obj-$(subst m,y,$(CONFIG_SERIAL_MRST_MAX3110)) += platform_max3111.o
 #obj-$(subst m,y,$(CONFIG_KEYBOARD_GPIO)) += platform_gpio_keys.o
 # WIFI devices
 obj-$(subst m,y,$(CONFIG_MMC_SDHCI_ACPI)) += platform_sdio_regulator.o
+obj-y += platform_iwlwifi.o
diff --git a/arch/x86/platform/intel-mid/device_libs/platform_iwlwifi.c b/arch/x86/platform/intel-mid/device_libs/platform_iwlwifi.c
new file mode 100644
index 0000000..a6f0e15
--- /dev/null
+++ b/arch/x86/platform/intel-mid/device_libs/platform_iwlwifi.c
@@ -0,0 +1,108 @@
+/*
+ * platform_iwlwifi_pcie.c: iwlwifi pcie platform device.
+ *
+ * (C) Copyright 2014 Intel Corporation
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; version 2
+ * of the License.
+ */
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/gpio.h>
+#include <linux/pci.h>
+#include <linux/delay.h>
+#include <linux/platform_device.h>
+#include <linux/platform_data/iwlwifi.h>
+#ifdef CONFIG_ACPI
+#include <linux/acpi.h>
+#endif
+
+
+#ifdef CONFIG_ACPI
+static struct acpi_device_id iwlwifi_acpi_id[] = {
+	{ "INT5502" , 0 },
+	{ "INT8260" , 0 },
+	{},
+};
+
+static int iwlwifi_acpi_match_platform(struct device *dev, void *data)
+{
+	struct acpi_device_id *ids = data, *id;
+	struct platform_device *pdev = to_platform_device(dev);
+
+	if (!pdev)
+		return FALSE;
+
+	for (id = ids; id->id[0]; id++) {
+		if (!strncmp(id->id, pdev->name, strlen(id->id)))
+			return TRUE;
+	}
+	return FALSE;
+}
+#endif
+
+static void iwlwifi_pcie_enable_regulator(void)
+{
+#ifdef CONFIG_ACPI
+	struct device *dev = NULL;
+	struct acpi_device *adev = NULL;
+	acpi_handle handle;
+
+	dev = bus_find_device(&platform_bus_type, NULL, iwlwifi_acpi_id,
+			      iwlwifi_acpi_match_platform);
+
+	if (dev && ACPI_HANDLE(dev)) {
+		handle = ACPI_HANDLE(dev);
+		if (acpi_bus_get_device(handle, &adev) == 0)
+			acpi_device_set_power(adev, ACPI_STATE_D0);
+	} else
+		pr_err("%s device not found\n", __func__);
+#endif
+	msleep(100);
+
+}
+
+static void iwlwifi_pcie_disable_regulator(void)
+{
+#ifdef CONFIG_ACPI
+	struct device *dev = NULL;
+	struct acpi_device *adev = NULL;
+	acpi_handle handle;
+
+	dev = bus_find_device(&platform_bus_type, NULL, iwlwifi_acpi_id,
+			      iwlwifi_acpi_match_platform);
+
+	if (dev && ACPI_HANDLE(dev)) {
+		handle = ACPI_HANDLE(dev);
+		if (acpi_bus_get_device(handle, &adev) == 0)
+			acpi_device_set_power(adev, ACPI_STATE_D3_COLD);
+	} else
+		pr_err("%s device not found\n", __func__);
+#endif
+}
+
+static struct iwl_trans_platform_ops iwlwifi_pcie_platform_ops = {
+	iwlwifi_pcie_enable_regulator,
+	iwlwifi_pcie_disable_regulator
+};
+
+static struct platform_device iwlwifi_pcie_platform_device = {
+	.name = "iwlwifi_pcie_platform",
+	.id = -1,
+	.dev = {
+		.platform_data = &iwlwifi_pcie_platform_ops,
+	},
+};
+
+int __init iwlwifi_platform_init(void)
+{
+	int ret = 0;
+
+	ret = platform_device_register(&iwlwifi_pcie_platform_device);
+	pr_info("iwlwifi platform init\n");
+
+	return ret;
+}
+device_initcall(iwlwifi_platform_init);
diff --git a/drivers/acpi/acpi_platform.c b/drivers/acpi/acpi_platform.c
index 9064648..c0e03b9 100644
--- a/drivers/acpi/acpi_platform.c
+++ b/drivers/acpi/acpi_platform.c
@@ -43,6 +43,8 @@ static const struct acpi_device_id acpi_platform_device_ids[] = {
 	{ "BCM4752E" },
 	{ "BCM43241" },
 	{ "BCM4356" },
+	{ "INT5502"  },
+	{ "INT8260"  },
 	{ "BCM47521" },
 	/* Modem Controller Device */
 	{ "MCD0001", 0 },
diff --git a/include/linux/platform_data/iwlwifi.h b/include/linux/platform_data/iwlwifi.h
new file mode 100644
index 0000000..8142ec4
--- /dev/null
+++ b/include/linux/platform_data/iwlwifi.h
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2015 Intel Corporation
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+ * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifndef _LINUX_IWLWIFI_PLATFORM_H
+#define _LINUX_IWLWIFI_PLATFORM_H
+
+#define IWLWIFI_PLATFORM_NAME "iwlwifi_pcie_platform"
+
+/**
+ * struct iwl_trans_platform_ops - platform specific hooks.
+ *
+ * @enable_regulator: enable the regulator supplying power to the card.
+ *	May sleep
+ * @disable_regulator: shutting down the main regulator attached to the card.
+ *	May sleep
+ */
+struct iwl_trans_platform_ops {
+	void (*enable_regulator)(void);
+	void (*disable_regulator)(void);
+};
+
+#endif /* _LINUX_IWLWIFI_PLATFORM_H */
-- 
1.9.1

